### ISSUE TITLE
extension viewcode fails with function aliases

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#623]
My Python package (http://github.com/AndreaCensi/geometry) has the following structure. 

I divide the implementation of functions in various modules, but then I import the functions in the main module's **init**. In this way, the user sees a cleaner interface.

```
+ geometry/

  + module1.py

    def func1(): ...
    def func2(): ...

  + module2.py

    def func3(): ...

  + __init__.py

    from .module1 import func1,func2
    from .module2 import func3
```

The user see this interface:

```
#!python

from geometry import func1, func3
```

However, when I document the code using this sphinx syntax:

```
.. py:module:: geometry

.. autofunction:: func1
.. autofunction:: func2
.. autofunction:: func3
```

Sphinx does find the docstring for the functions, but the viewcode extension cannot find the code.

If I use this:

```
.. py:module:: geometry

.. autofunction:: geometry.module1.func1
.. autofunction:: geometry.module1.func2
.. autofunction:: geometry.module2.func3
```

then everything works. However this solution exposes the internal implementation details (subdivision among module1,module2) which I want to remain hidden from the user.

The following are links to the actual code.

See example sphinx markup here:

https://github.com/AndreaCensi/geometry/raw/master/docs/source/api.rst

and the resulting HTML which doesn't show the viewcode links here:

http://andreacensi.github.com/geometry/api.html

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/623
- Originally reported by: Andrea Censi
- Originally created at: 2011-01-29T07:41:12.733

**Comment by shimizukawa** [comment]
_From Andrea Censi on 2011-01-29 07:47:23+00:00_

I can confirm that this happens even with 1.1pre.

**Comment by shimizukawa** [comment]
_From Ali-Akber  Saifee on 2011-07-01 08:51:56+00:00_

I'm not sure if this is due to the exact same problem, but my project is also unable to show the 'view source' links, while the docstrings are correctly picked up.
The module structure is:
{{{
module
 |- sub_module
      |-- **init**.py
      |-- package_one.py
      |-- package_two.py
}}}

Where **init**.py:
{{{
#!python

from package_one import ClassA, ClassB
from package_one import funcA, funcB
**all** = ["ClassA","ClassB", "funcA", "funcB"]
}}}

package_one.py:
{{{
#!python
class A:
    """
    I am A
    """
    pass

class B:
    """
    i am B
    """
    pass
}}}

and package_two.py
{{{
#!python
def funcA():
    """
    I am A too
    """
    pass

def funcB():
    """
    I am B too
    """
    pass

}}}

The sphinx source for the autodoc is:
{{{
.. automodule :: module.sub_module
    :members:
}}}

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2011-09-22 10:30:05+00:00_

This cannot work today because of the way viewcode gets at the information about an object's module. I'll keep it open as a feature request though.

**Comment by shimizukawa** [comment]
_From Anonymous on 2011-11-30 11:57:19+00:00_

I've made a rather messy patch that might give a starting point for fixing this. It uses autodoc's autodoc-process-docstring signal (which usefully passes an object that you can get a **module** attribute from) and puts it into an internal lookup table. I'm not familiar with how sphinx works internally so apologies if this isn't the right way to do things.
{{{
diff -ur Sphinx-1.1_original/sphinx/ext/viewcode.py Sphinx-1.1/sphinx/ext/viewcode.py
--- Sphinx-1.1_original/sphinx/ext/viewcode.py  2011-11-30 10:07:31.000000000 +0000
+++ Sphinx-1.1/sphinx/ext/viewcode.py   2011-11-30 11:45:08.000000000 +0000
@@ -55,6 +55,9 @@
             if not modname:
                 continue
             fullname = signode.get('fullname')
-            lookupname = u"%s.%s" %(modname, fullname)
-            if lookupname in env._viewcode_realmodules:
- ```
             modname = env._viewcode_realmodules[lookupname]            
       if not has_tag(modname, fullname, env.docname):
           continue
       if fullname in names:
  ```
  
  @@ -168,10 +171,19 @@
  
   yield ('_modules/index', context, 'page.html')

+# horrible hack that piggy backs on autodoc's object finding system and makes
+# a lookup of documentation name -> real module name so that we can find the
+# sourcecode even if the module does some odd environment things
+def process_docstring(app, objtype, fullname, ob, options, docstringlines):
-    if not hasattr(app.builder.env, '_viewcode_realmodules'):
-        app.builder.env._viewcode_realmodules = {}
-    if getattr(ob, "**module**", None):
- ```
     app.builder.env._viewcode_realmodules[fullname] = getattr(ob, "__module__") 
  ```
  
  def setup(app):
     app.connect('doctree-read', doctree_read)
     app.connect('html-collect-pages', collect_pages)
     app.connect('missing-reference', missing_reference)
-    app.connect('autodoc-process-docstring', process_docstring)
   #app.add_config_value('viewcode_include_modules', [], 'env')
   #app.add_config_value('viewcode_exclude_modules', [], 'env')
  }}}

**Comment by shimizukawa** [comment]
_From [t_only_ttt](https://bitbucket.org/t_only_ttt) on 2011-12-02 14:21:43+00:00_

Oops, that should be:

{{{
diff -ur Sphinx-1.1_original/sphinx/ext/viewcode.py Sphinx-1.1/sphinx/ext/viewcode.py
--- Sphinx-1.1_original/sphinx/ext/viewcode.py  2011-11-30 10:07:31.000000000 +0000
+++ Sphinx-1.1/sphinx/ext/viewcode.py   2011-11-30 11:45:08.000000000 +0000
@@ -55,6 +55,9 @@
             if not modname:
                 continue
             fullname = signode.get('fullname')
-            lookupname = u"%s.%s" %(modname, fullname)
-            if lookupname in getattr(env, "_viewcode_realmodules", []):
- ```
             modname = env._viewcode_realmodules[lookupname]            
       if not has_tag(modname, fullname, env.docname):
           continue
       if fullname in names:
  ```
  
  @@ -168,10 +171,19 @@
  
   yield ('_modules/index', context, 'page.html')

+# horrible hack that piggy backs on autodoc's object finding system and makes
+# a lookup of documentation name -> real module name so that we can find the
+# sourcecode even if the module does some odd environment things
+def process_docstring(app, objtype, fullname, ob, options, docstringlines):
-    if not hasattr(app.builder.env, '_viewcode_realmodules'):
-        app.builder.env._viewcode_realmodules = {}
-    if getattr(ob, "**module**", None):
- ```
     app.builder.env._viewcode_realmodules[fullname] = getattr(ob, "__module__") 
  ```
  
  def setup(app):
     app.connect('doctree-read', doctree_read)
     app.connect('html-collect-pages', collect_pages)
     app.connect('missing-reference', missing_reference)
-    app.connect('autodoc-process-docstring', process_docstring)
   #app.add_config_value('viewcode_include_modules', [], 'env')
   #app.add_config_value('viewcode_exclude_modules', [], 'env')

}}}

**Comment by shimizukawa** [comment]
_From Anonymous on 2011-12-05 14:25:51+00:00_

An improved patch for fixing this issue which I think fixes it more correctly:

{{{
--- Sphinx-1.1/sphinx/ext/viewcode.py   2011-10-06 10:07:32.000000000 +0100
+++ Sphinx-1.1/sphinx/ext/viewcode.py   2011-12-05 13:44:24.000000000 +0000
@@ -16,13 +16,41 @@
 from sphinx.pycode import ModuleAnalyzer
 from sphinx.util.nodes import make_refnode

+import sys
+import traceback
+

 def doctree_read(app, doctree):
     env = app.builder.env
     if not hasattr(env, '_viewcode_modules'):
         env._viewcode_modules = {}
-    def has_tag(modname, fullname, docname):
-    def get_full_modname(modname, attribute):
-        try:
-            **import**(modname)
-        except Exception, error:
-            if not app.quiet:
-                app.info(traceback.format_exc().rstrip())
-            app.warn('viewcode can\'t import %s, failed with error "%s"' %
-                (modname, error))
-            return None
-        module = sys.modules[modname]
-        try:
-            # Allow an attribute to have multiple parts and incidentially allow
-            # repeated .s in the attribute.
-            attr = attribute.split('.')
-            value = module
-            for attr in attribute.split('.'):
-                if attr:
-                    value = getattr(value, attr)
-        except AttributeError:
-            app.warn('Didn\'t find %s in %s' % (attribute, module.**name**))
-            return None
-        else:
-            return getattr(value, '**module**', None)
  +
  +
-    def has_tag(modname, fullname, docname, refname):
       entry = env._viewcode_modules.get(modname, None)
       if entry is None:
           try:
  @@ -35,15 +63,16 @@
               code = analyzer.code.decode(analyzer.encoding)
           else:
               code = analyzer.code
-            entry = code, analyzer.tags, {}
-            entry = code, analyzer.tags, {}, refname
           env._viewcode_modules[modname] = entry
       elif entry is False:
           return
-        code, tags, used = entry
-        _, tags, used, _ = entry
       if fullname in tags:
           used[fullname] = docname
           return True

+
     for objnode in doctree.traverse(addnodes.desc):
         if objnode.get('domain') != 'py':
             continue
@@ -52,10 +81,13 @@
             if not isinstance(signode, addnodes.desc_signature):
                 continue
             modname = signode.get('module')
-            fullname = signode.get('fullname')
-            refname = modname
-            if env.config.viewcode_import:
-                modname = get_full_modname(modname, fullname)
           if not modname:
               continue
-            fullname = signode.get('fullname')
-            if not has_tag(modname, fullname, env.docname):
-            if not has_tag(modname, fullname, env.docname, refname):
               continue
           if fullname in names:
               # only one link per name, please
  @@ -94,7 +126,7 @@
   for modname, entry in env._viewcode_modules.iteritems():
       if not entry:
           continue
-        code, tags, used = entry
-        code, tags, used, refname = entry
       # construct a page name for the highlighted source
       pagename = '_modules/' + modname.replace('.', '/')
       # highlight the source using the builder's highlighter
  @@ -111,7 +143,7 @@
       maxindex = len(lines) - 1
       for name, docname in used.iteritems():
           type, start, end = tags[name]
-            backlink = urito(pagename, docname) + '#' + modname + '.' + name
- ```
         backlink = urito(pagename, docname) + '#' + refname + '.' + name
       lines[start] = (
           '<div class="viewcode-block" id="%s"><a class="viewcode-back" '
           'href="%s">%s</a>' % (name, backlink, _('[docs]'))
  ```
  
  @@ -170,6 +202,7 @@
  
  def setup(app):
-    app.add_config_value('viewcode_import', True, False)
   app.connect('doctree-read', doctree_read)
   app.connect('html-collect-pages', collect_pages)
   app.connect('missing-reference', missing_reference)

}}}

**Comment by shimizukawa** [comment]
_From [t_only_ttt](https://bitbucket.org/t_only_ttt) on 2011-12-05 16:48:53+00:00_

I agree, that looks a lot cleaner and doesn't rely on signals from autodoc...

**Comment by shimizukawa** [comment]
_From Werner Van Geit on 2013-04-17 12:04:33+00:00_

Is there any timeline for when these patches would be incorporated in the code ? I have exactly the same issue.

**Comment by shimizukawa** [comment]
_From [dolfim](https://bitbucket.org/dolfim) on 2013-09-10 12:47:04+00:00_

I found the above patch very helpful. I created a pull request that hopefully could speed up a bit the fix.
Pull request #163

**Comment by shimizukawa** [comment]
_From [mattions](https://bitbucket.org/mattions) on 2013-10-09 15:08:36+00:00_

Got same issue here.

Any news on this front?

**Comment by shimizukawa** [comment]
_From subhasis ray on 2014-08-21 10:14:18+00:00_

bump!

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-08-22 03:24:07+00:00_

:mod:`~sphinx.ext.viewcode` support imported function/class aliases. Closes #623

→ <<cset e8b870de0cbaf724986b65de9552262ac253a6d2>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-08-22 03:27:37+00:00_

Done!
